unit_tests/wind_energy/test_abl_src.cpp

```diff
@@ -282,48 +223,107 @@ void init_abl_temperature_field(int kdim,
 
 }
 
-TEST_F(ABLTest, boussinesq)
+TEST_F(ABLMeshTest, boussinesq)
 {
-    constexpr int kdim = 16;
+    constexpr int kdim = 7;
     constexpr amrex::Real tol = 1.0e-12;
 
     // Initialize parameters
     utils::populate_abl_params();
+    initialize_mesh();
+
+    auto& pde_mgr = sim().pde_manager();
+    pde_mgr.register_icns();
+    pde_mgr.register_transport_pde("Temperature");
+    sim().init_physics();
+
+    amr_wind::pde::icns::BoussinesqBuoyancy bb(sim());
+
+    auto& src_term = pde_mgr.icns().fields().src_term;
+
+    auto& temperature = sim().repo().get_field("temperature",amr_wind::FieldState::Old);
+
+    src_term.setVal(0.0);
 
-    // Create velocity and source terms fields
-    amrex::Box bx{{0, 0, 0}, {2, 2, kdim}};
-    amrex::FArrayBox temperature(bx, 1);
-    amrex::FArrayBox vel_src(bx, AMREX_SPACEDIM);
+    run_algorithm(temperature, [&](const int lev, const amrex::MFIter& mfi) {
 
-    amr_wind::BoussinesqBuoyancyOld bb;
+        const auto bx = mfi.validbox();
+        const auto& temp_arr = temperature(lev).array(mfi);
+        const auto& src_arr = src_term(lev).array(mfi);
 
-    temperature.setVal<amrex::RunOn::Device>(0.0);
-    vel_src.setVal<amrex::RunOn::Device>(0.0);
-    init_abl_temperature_field(kdim, bx, temperature);
+        init_abl_temperature_field(kdim, bx, temp_arr);
+        bb(lev, mfi, bx, amr_wind::FieldState::Old, src_arr);
 
-    bb(bx, temperature.array(), vel_src.array());
+    });
 
     // should be no forcing in x and y directions
-    EXPECT_NEAR(vel_src.min<amrex::RunOn::Device>(0), 0.0, tol);
-    EXPECT_NEAR(vel_src.max<amrex::RunOn::Device>(0), 0.0, tol);
-    EXPECT_NEAR(vel_src.min<amrex::RunOn::Device>(1), 0.0, tol);
-    EXPECT_NEAR(vel_src.max<amrex::RunOn::Device>(1), 0.0, tol);
+    for (int i=0; i < 2; ++i) {
+        const auto min_src = utils::field_min(src_term, i);
+        const auto max_src = utils::field_max(src_term, i);
+        EXPECT_NEAR(min_src, 0.0, tol);
+        EXPECT_NEAR(max_src, 0.0, tol);
+    }
 
-//    f = beta * (T0 - T)*g
-    EXPECT_NEAR(vel_src.min<amrex::RunOn::Device>(2), 0.0, tol);
-    EXPECT_NEAR(vel_src.max<amrex::RunOn::Device>(2), -9.81*(300.0-308.0)/300.0, tol);
+    // f = beta * (T0 - T)*g
+    EXPECT_NEAR(utils::field_min(src_term, 2), 0.0, tol);
+    EXPECT_NEAR(utils::field_max(src_term, 2), -9.81*(300.0-308.0)/300.0, tol);
 
 }
 
+TEST_F(ABLMeshTest, boussinesq_nph)
+{
+    constexpr int kdim = 7;
+    constexpr amrex::Real tol = 1.0e-12;
+
+    // Initialize parameters
+    utils::populate_abl_params();
+    initialize_mesh();
+
+    auto& pde_mgr = sim().pde_manager();
+    pde_mgr.register_icns();
+    pde_mgr.register_transport_pde("Temperature");
+    sim().init_physics();
+
+    amr_wind::pde::icns::BoussinesqBuoyancy bb(sim());
+
+    auto& src_term = pde_mgr.icns().fields().src_term;
+
+    auto& temperature = sim().repo().get_field("temperature").create_state(amr_wind::FieldState::NPH);
```

Comment:
This is not needed anymore. `register_transport_pde("Temperature")` already creates `n + 1/2` states for you.